core/types/block_test.go

```diff
@@ -423,3 +428,83 @@ func TestCheckTransactionConditional(t *testing.T) {
 		})
 	}
 }
+
+func getTestHeader() *Header {
+	return &Header{
+		ParentHash:      common.HexToHash("0x112233445566778899001122334455667788990011223344556677889900aabb"),
+		UncleHash:       common.HexToHash("0x2233445566778899001122334455667788990011223344556677889900aabbcc"),
+		Coinbase:        common.HexToAddress("0x8888f1f195afa192cfee860698584c030f4c9db1"),
+		Root:            common.HexToHash("0x33445566778899001122334455667788990011223344556677889900aabbccdd"),
+		TxHash:          common.HexToHash("0x445566778899001122334455667788990011223344556677889900aabbccddee"),
+		ReceiptHash:     common.HexToHash("0x5566778899001122334455667788990011223344556677889900aabbccddeeff"),
+		WithdrawalsHash: &EmptyWithdrawalsHash,
+		Bloom:           Bloom{},
+		Difficulty:      big.NewInt(131072),
+		Number:          big.NewInt(42),
+		GasLimit:        3141592,
+		GasUsed:         21000,
+		Time:            1426516743,
+		Extra:           []byte("extra data"),
+		MixDigest:       common.HexToHash("0x66778899001122334455667788990011223344556677889900aabbccddeeff00"),
+		Nonce:           EncodeNonce(0xa13a5a8c8f2bb1c4),
+		BaseFee:         big.NewInt(1000000000),
+	}
+}
+
+func checkFields(t *testing.T, header, want *Header) {
+	check := func(f string, got, want interface{}) {
+		if !reflect.DeepEqual(got, want) {
+			t.Errorf("%s mismatch: got %v, want %v", f, got, want)
+		}
+	}
+
+	check("ParentHash", header.ParentHash, want.ParentHash)
+	check("UncleHash", header.UncleHash, want.UncleHash)
+	check("Coinbase", header.Coinbase, want.Coinbase)
+	check("Root", header.Root, want.Root)
+	check("TxHash", header.TxHash, want.TxHash)
+	check("ReceiptHash", header.ReceiptHash, want.ReceiptHash)
+	check("WithdrawalsHash", header.WithdrawalsHash, want.WithdrawalsHash)
+	check("Bloom", header.Bloom, want.Bloom)
+	check("Difficulty", header.Difficulty, want.Difficulty)
+	check("Number", header.Number, want.Number)
+	check("GasLimit", header.GasLimit, want.GasLimit)
+	check("GasUsed", header.GasUsed, want.GasUsed)
+	check("Time", header.Time, want.Time)
+	check("Extra", header.Extra, want.Extra)
+	check("MixDigest", header.MixDigest, want.MixDigest)
+	check("Nonce", header.Nonce, want.Nonce)
+	check("BaseFee", header.BaseFee, want.BaseFee)
+}
+
+func TestHeaderRLPEncoding(t *testing.T) {
+	header := getTestHeader()
+
+	encoded, err := rlp.EncodeToBytes(header)
+	if err != nil {
+		t.Fatalf("failed to encode header: %v", err)
+	}
+
+	var decoded Header
+	if err := rlp.DecodeBytes(encoded, &decoded); err != nil {
+		t.Fatalf("failed to decode header: %v", err)
+	}
+
+	checkFields(t, &decoded, header)
+}
+
+func TestHeaderJSONEncoding(t *testing.T) {
```

Comment:
I didn't see this covered already when looking around the source code, I wish we had better test coverage. It could make sense to turn this into a fuzz test rather than hardcoding the test header, similar to https://github.com/ethereum/go-ethereum/blob/aa636921291744af082cefbe4dd811fb58fbde63/core/types/deposit_test.go#L40